### ISSUE TITLE
Add support for fmt.GoStringer 

### ISSUE THREAD

**Issue by a69** [shopspring/decimal#270]
we  need to print Go-syntax representation of decimal variables.
```
package main
import (
	"fmt"

	"github.com/shopspring/decimal"
)

func main() {
	one := decimal.NewFromInt(1)

	fmt.Printf("%#v\n", one)

	//print: 	decimal.Decimal{value:(*big.Int)(0xc00011e3a0), exp:0}
	//expected:	decimal.NewFromString("1")
}

**Comment by mwoss** [comment]
Hi! Personally, I don't think such change is anyhow valuable for the library. `GoString` method should defines the Go syntax for that value and the default value is more valuable in my opinion than the proposed solution. Returning the `NewFromString` init method in such a case makes no sense. I would suggest forking the library or implementing a custom GoString method for a decimal struct in your own project.
The only meaningful change I can see for GoString method would be printing the value of `decimal.Value` instead of just a memory address for a pointer. Something like this:
```go
// before: 
decimal.Decimal{value:(*big.Int)(0xc00011e3a0), exp:0}

//after (first or second approach)
decimal.Decimal{value:1, exp:0} 
decimal.Decimal{value:1 (*big.Int)(0xc00011e3a0), exp:0}
```

**Comment by a69** [comment]
First, we use GoStringer interface for generate data in go source code for Unit Test from actual values. Right now we have to define GoString, Add, Div and many other method to use our custom decimal struct in our project. I believe this change help people to generate easily Go code with this package. 
Second, your proposal change is not valid go syntax, you set private fields outside of package!
For example you can see the implementation of GoString in time package. https://github.com/golang/go/blob/master/src/time/format.go#L528

**Comment by mwoss** [comment]
That's true if GoString must print valid syntax that can be used outside of the package my proposal is not valid at all, I was not familiar with this interface beforehand.
On the other hand printing value syntax using one of few existing init methods feels odd, it's not generic enough and could be confusing for other users of the library. Initialization via struct literal would be the best here, but not possible as you mentioned earlier.
Additionally the problem with `decimal.RequireFromString(" + d.String() + ")` is that we might lose precision while casting to string, all trailing zeroes are removed.  This is something unacceptable imo.

**Comment by a69** [comment]
I think if `big.Int` has a Method like decimal has `func RequireFromString(value string) Decimal` something like: 
`func (z *Int) RequireSetString(s string, base int) *Int` we could a painless implementation like:
`decimal.NewFromBigInt(new(big.Int).RequireSetString("+d.value.string+",10)),"+fmt.Sprin(d.exp)+")`.  but as there is no such function in `big.Int` and work with `SetBits` and `SetBytes` is more complex and output is less human-readable.I think rightnow best solution is  use somethig like `decimal.RequireFromString(" + d.string(false) + ")` to prevent  lose precision

**Comment by treuherz** [comment]
Just wanted to add 2¢, which is that the current default behaviour of printing a memory address for the int is really unhelpful for most cases. I can see that making sure you've not accidentally copied a pointer or something is important for developing the package, but as a user of it the chances for that happening are pretty slim, but needing to see what the value is is pretty common.

In our use case, the problem comes up when using [`testing/quick`](https://pkg.go.dev/testing/quick). The [way it formats test failures](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/testing/quick/quick.go;l=382;drc=02bd0fd39cb6ffd840901ea751a61047ccad47cb) uses `%#v` with no way to change it, which is essentially completely opaque.

I've left more comments on what output would be most helpful on the PR, but briefly I think returning syntax using `decimal.New` is more helpful than `decimal.RequireFromString`.

**Comment by mwoss** [comment]
Sorry for not being active lately, started a new job. 
My biggest concern about the proposed enhancement/PR is that decimal returned by `GoString` method may not always reflect the original decimal due to potential loss of precision - `decimal.String()` removes all trailing 0s.
As one of the main goals of this library is precision and correctness, I am not entirely convinced about the proposed solution. 

Btw, did you consider initialization via NewBigInt? We can retrieve copies of decimal value and exponent via `Expontent` and `Coefficient` methods.
